app/gui/src/project-view/util/codemirror/persistence/scroll.ts

```diff
@@ -0,0 +1,149 @@
+  margin: z.number(),
+})
+const scrollSchema = z.object({
+  range: z.object({
+    anchor: z.number(),
+    head: z.number().optional(),
+  }),
+  x: scrollAxisSchema.optional(),
+  y: scrollAxisSchema.optional(),
+})
+
+export type ScrollState = z.infer<typeof scrollSchema>
+
+type ScrollTarget = ReturnType<EditorView['scrollSnapshot']>['value']
+const scrollRestoreStateEffect = StateEffect.define()
+
+export interface ScrollStatePluginOptions {
+  /** Whether to persist the scroll position along the y-axis. */
+  y: boolean
+  /**
+   * Whether to persist the scroll position along the x-axis. Note that, if the view is not
+   * scrollable on the x-axis, this MUST be `false`. If it is `true` but the editor's element is not
+   * horizontally scrollable, restoring state may result in scrolling an ancestor of the editor.
+   */
+  x: boolean
+}
+
+/** {@link PersistableStatePlugin} for editor scroll position. */
+class ScrollStatePlugin implements PluginValue, PersistableStatePlugin<ScrollState> {
+  private readonly scrollEffectRef: ShallowRef<undefined> = shallowRef()
+  private restoringScroll: boolean = false
+  private readonly x: boolean
+  private readonly y: boolean
+  constructor(
+    private readonly view: EditorView,
+    { x, y }: ScrollStatePluginOptions,
+  ) {
+    this.x = x
+    this.y = y
+  }
+
+  update(update: ViewUpdate) {
+    if (
+      (update.docChanged || update.viewportChanged) &&
+      !update.transactions.every((tr) => tr.effects.some((e) => e.is(scrollRestoreStateEffect)))
+    )
+      this.scrolled()
+  }
+
+  onScroll() {
+    this.scrolled()
+  }
+
+  private scrolled() {
+    if (!this.restoringScroll) triggerRef(this.scrollEffectRef)
+  }
+
+  captureState(): ScrollState {
+    const _scrollEffect = this.scrollEffectRef.value
+    return this.serializeState(this.view.scrollSnapshot().value)
+  }
+
+  private serializeState(snapshot: ScrollTarget): ScrollState {
+    const { range, y, x, yMargin, xMargin } = snapshot
+    return {
+      range: {
+        anchor: range.anchor,
+        head: range.head,
+      },
+      ...(this.y ?
+        {
+          y: {
+            mode: y,
+            margin: yMargin,
+          },
+        }
+      : {}),
+      ...(this.x ?
+        {
+          x: {
+            mode: x,
+            margin: xMargin,
+          },
+        }
+      : {}),
+    }
+  }
+
+  restoreState(rawState: unknown) {
+    const state = ScrollStatePlugin.parseState(rawState)
+    if (!state) return
+    const { range, y, x } = state
+    this.restoringScroll = true
+    try {
+      const selectionRange = EditorSelection.range(range.anchor, range.head ?? range.anchor)
+      const scrollOptions = {
+        isSnapshot: true,
+        ...(this.y && y ?
+          {
+            y: y.mode,
+            yMargin: y.margin,
+          }
+        : {}),
+        ...(this.x && x ?
+          {
+            x: x.mode,
+            xMargin: x.margin,
+          }
+        : {}),
```

Comment:
Similariliy here, I don't think we care about those options having keys with `undefined` values.